File: src/timing.py
from __future__ import annotations

import time
from typing import Callable, Optional


def wait_until(
    pred: Callable[[], bool],
    timeout_ms: int,
    poll_ms: int = 4,
    should_stop: Optional[Callable[[], bool]] = None,
) -> bool:
    """条件成立立刻返回 True；超时或停止返回 False。"""
    deadline = time.monotonic() + max(0, timeout_ms) / 1000.0
    interval = max(0.001, poll_ms / 1000.0)
    while True:
        if should_stop and should_stop():
            return False
        if pred():
            return True
        remain = deadline - time.monotonic()
        if remain <= 0:
            return False
        time.sleep(min(interval, remain))

File: electron/engine/timing.ts
export async function sleepMs(ms: number, shouldStop?: () => boolean): Promise<boolean> {
  if (ms <= 0) return Boolean(shouldStop?.());
  const end = Date.now() + ms;
  while (Date.now() < end) {
    if (shouldStop?.()) return true;
    await new Promise((r) => setTimeout(r, Math.min(50, Math.max(0, end - Date.now()))));
  }
  return Boolean(shouldStop?.());
}

export async function waitUntil(
  pred: () => boolean | Promise<boolean>,
  timeoutMs: number,
  pollMs = 4,
  shouldStop?: () => boolean,
): Promise<boolean> {
  const deadline = Date.now() + Math.max(0, timeoutMs);
  const interval = Math.max(1, pollMs);
  while (true) {
    if (shouldStop?.()) return false;
    if (await pred()) return true;
    const remain = deadline - Date.now();
    if (remain <= 0) return false;
    await new Promise((r) => setTimeout(r, Math.min(interval, remain)));
  }
}
